src/graph/sampler.cc

```diff
@@ -0,0 +1,463 @@
+    // We need to avoid resampling the same node in a layer, but we allow a node
+    // to be resampled in multiple layers. We use `sub_ver_map` to keep track of
+    // sampled nodes in a layer, and clear it when entering a new layer.
+    sub_ver_map.clear();
+    // sub_vers is used both as a node collection and a queue.
+    // In the while loop, we iterate over sub_vers and new nodes are added to the vector.
+    // A vertex in the vector only needs to be accessed once. If there is a vertex behind idx
+    // isn't in the last level, we will sample its neighbors. If not, the while loop terminates.
+    while (idx < sub_vers.size() && layer_id - 1 == sub_vers[idx].second) {
+      dgl_id_t dst_id = sub_vers[idx].first;
+      int cur_node_level = sub_vers[idx].second;
+      idx++;
+
+      tmp_sampled_src_list.clear();
+      tmp_sampled_edge_list.clear();
+      dgl_id_t ver_len = *(indptr+dst_id+1) - *(indptr+dst_id);
+      if (probability == nullptr) {  // uniform-sample
+        GetUniformSample(val_list + *(indptr + dst_id),
+                         col_list + *(indptr + dst_id),
+                         ver_len,
+                         num_neighbor,
+                         &tmp_sampled_src_list,
+                         &tmp_sampled_edge_list,
+                         &time_seed);
+      } else {  // non-uniform-sample
+        GetNonUniformSample(probability,
+                            val_list + *(indptr + dst_id),
+                            col_list + *(indptr + dst_id),
+                            ver_len,
+                            num_neighbor,
+                            &tmp_sampled_src_list,
+                            &tmp_sampled_edge_list,
+                            &time_seed);
+      }
+      CHECK_EQ(tmp_sampled_src_list.size(), tmp_sampled_edge_list.size());
+      size_t pos = neighbor_list.size();
+      neigh_pos.emplace_back(dst_id, pos);
+      // First we push the size of neighbor vector
+      neighbor_list.push_back(tmp_sampled_edge_list.size());
+      // Then push the vertices
+      for (size_t i = 0; i < tmp_sampled_src_list.size(); ++i) {
+        neighbor_list.push_back(tmp_sampled_src_list[i]);
+      }
+      // Finally we push the edge list
+      for (size_t i = 0; i < tmp_sampled_edge_list.size(); ++i) {
+        neighbor_list.push_back(tmp_sampled_edge_list[i]);
+      }
+      num_edges += tmp_sampled_src_list.size();
+      for (size_t i = 0; i < tmp_sampled_src_list.size(); ++i) {
+        // We need to add the neighbor in the hashtable here. This ensures that
+        // the vertex in the queue is unique. If we see a vertex before, we don't
+        // need to add it to the queue again.
+        auto ret = sub_ver_map.insert(tmp_sampled_src_list[i]);
+        // If the sampled neighbor is inserted to the map successfully.
+        if (ret.second) {
+          sub_vers.emplace_back(tmp_sampled_src_list[i], cur_node_level + 1);
+        }
+      }
```

Comment:
Give `num_neighbor=10` and a node have 100 neighbors, does this means that it is possible no neighbor is sampled for that nodes if they *all* happen to be picked by other nodes before?

Reply:
i don't think it's possible. all sampled neighbors will be stored.

Reply:
sub_vers stores the nodes whose neighbors will be sampled in the next layer.